segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLogFactory.java

```diff
@@ -169,5 +193,74 @@ private BookKeeper startBookKeeperClient() throws Exception {
                          .build();
     }
 
+    /**
+     * Recreate the Bookkeeper client if a given log exhibits MAX_CREATE_ATTEMPTS_PER_LOG creation attempts (as a proxy
+     * for Container recoveries) within the period of time defined in LOG_CREATION_INSPECTION_PERIOD.
+     *
+     * @param logId Id of the log being restarted.
+     */
+    private void tryResetBookkeeperClient(int logId) {
+        synchronized (this) {
+            if (logInitializationTracker.containsKey(logId)) {
+                // Account for a restart of the Bookkeeper log.
+                logInitializationTracker.get(logId).incrementLogCreations();
+                // If the number of restarts for a single container is meets the threshold, let's reset the BK client.
+                if (logInitializationTracker.get(logId).isBookkeeperClientResetNeeded()
+                        && lastBookkeeperClientReset.get().getElapsed().compareTo(LOG_CREATION_INSPECTION_PERIOD) > 0) {
+                    try {
+                        log.info("Start creating Bookkeeper client in reset.");
+                        BookKeeper newClient = startBookKeeperClient();
+                        // If we have been able to create a new client successfully, reset the current one and update timer.
+                        log.info("Successfully created new Bookkeeper client, setting it as the new one to use.");
+                        BookKeeper oldClient = this.bookKeeper.getAndSet(newClient);
+                        lastBookkeeperClientReset.set(new Timer());
+                        // Lastly, attempt to close the old client.
+                        log.info("Attempting to close old client.");
+                        oldClient.close();
+                    } catch (Exception e) {
+                        log.error("Failure resetting the Bookkeeper client: ", e);
```

Comment:
This will result in double logging. Whomever is invoking this method will absolutely log it as well.

Reply:
Removed first log line.